src/graphene.cpp

```diff
@@ -0,0 +1,1562 @@
+                          " %d size %d bytes\n",
+                inv.hash.ToString(), pfrom->GetLogName(), nHops, nSizeGrapheneBlock);
+            return true;
+        }
+
+        // Request full block if this one isn't extending the best chain
+        if (pIndex->nChainWork <= chainActive.Tip()->nChainWork)
+        {
+            std::vector<CInv> vGetData;
+            vGetData.push_back(inv);
+            pfrom->PushMessage(NetMsgType::GETDATA, vGetData);
+
+            graphenedata.ClearGrapheneBlockData(pfrom, grapheneBlock.header.GetHash());
+
+            LOGA("%s %s from peer %s received but does not extend longest chain; requesting full block\n", strCommand,
+                inv.hash.ToString(), pfrom->GetLogName());
+            return true;
+        }
+
+        {
+            LOG(GRAPHENE, "Received %s %s from peer %s. Size %d bytes.\n", strCommand, inv.hash.ToString(),
+                pfrom->GetLogName(), nSizeGrapheneBlock);
+
+            // Do not process unrequested grapheneblocks.
+            LOCK(pfrom->cs_mapgrapheneblocksinflight);
+            if (!pfrom->mapGrapheneBlocksInFlight.count(inv.hash))
+            {
+                dosMan.Misbehaving(pfrom, 10);
+                return error(
+                    "%s %s from peer %s but was unrequested\n", strCommand, inv.hash.ToString(), pfrom->GetLogName());
+            }
+        }
+    }
+
+    bool result = grapheneBlock.process(pfrom, nSizeGrapheneBlock, strCommand);
+
+    return result;
+}
+
+bool CGrapheneBlock::process(CNode *pfrom,
+    int nSizeGrapheneBlock,
+    std::string strCommand) // TODO: request from the "best" txn source not necessarily from the block source
+{
+    // In PV we must prevent two graphene blocks from simulaneously processing from that were recieved from the
+    // same peer. This would only happen as in the example of an expedited block coming in
+    // after an graphene request, because we would never explicitly request two graphene blocks from the same peer.
+    if (PV->IsAlreadyValidating(pfrom->id))
+        return false;
+
+    // Xpress Validation - only perform xval if the chaintip matches the last blockhash in the graphene block
+    bool fXVal;
+    {
+        LOCK(cs_main);
+        fXVal = (header.hashPrevBlock == chainActive.Tip()->GetBlockHash()) ? true : false;
+    }
+
+    graphenedata.ClearGrapheneBlockData(pfrom);
+    pfrom->nSizeGrapheneBlock = nSizeGrapheneBlock;
+
+    uint256 nullhash;
+    pfrom->grapheneBlock.nVersion = header.nVersion;
+    pfrom->grapheneBlock.nBits = header.nBits;
+    pfrom->grapheneBlock.nNonce = header.nNonce;
+    pfrom->grapheneBlock.nTime = header.nTime;
+    pfrom->grapheneBlock.hashMerkleRoot = header.hashMerkleRoot;
+    pfrom->grapheneBlock.hashPrevBlock = header.hashPrevBlock;
+    pfrom->grapheneBlockHashes.clear();
+    pfrom->grapheneBlockHashes.resize(nBlockTxs, nullhash);
```

Comment:
this resize makes me a little nervous. I'd prefer not to see it at all in any processing for p2p messages.  I understand what you're trying to do here performance wise but it's not worth it IMO.  If someone is somehow able to craft a fake graphene block that gives us a huge number for the vector.size() and subsequently the value of nBlockTxs  then we'd very quickly an out of memory condition.

Reply:
Done.